libcfnet/protocol.c

```diff
@@ -0,0 +1,88 @@
+/*
+   Copyright 2019 Northern.tech AS
+
+   This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+   This program is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by the
+   Free Software Foundation; version 3.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <protocol.h>
+
+#include <client_code.h>
+#include <client_protocol.h>
+#include <definitions.h>
+#include <net.h>
+
+
+Seq *CfNetOpenDir(AgentConnection *conn, char const *path)
+{
+    assert(conn != NULL);
+    assert(path != NULL);
+
+    char buf[CF_BUFSIZE] = {0};
+    snprintf(buf, CF_BUFSIZE, "OPENDIR %s", path);
+    size_t tosend = strlen(buf);
+
+    int ret = SendTransaction(conn->conn_info, buf, tosend, CF_DONE);
+    if (ret == -1)
+    {
+        return NULL;
+    }
+
+    Seq *seq = SeqNew(0, free);
+
+    bool more = true;
+    while (more)
+    {
+        ret = ReceiveTransaction(conn->conn_info, buf, (int *)&more);
+        if (ret == -1)
+        {
+            break;
+        }
+
+        if (FailedProtoReply(buf))
+        {
+            Log(LOG_LEVEL_INFO, "Network access to '%s:%s' denied",
+                conn->this_server, path);
+            break;
+        }
+
+        if (BadProtoReply(buf))
+        {
+            Log(LOG_LEVEL_INFO, "%s", buf + strlen("BAD: "));
+            break;
+        }
+
+        // FIXME Placeholder for the time being. There is probably a cleaner
+        //       and safer way to do this
+        for (int i = 0; buf[i] != '\0'; i += strlen(buf + i) + 1)
```

Comment:
Please explain the logic here, what does the response look like? Also, does ReceiveTransaction guarantee that buf is `\0` terminated, even for untrusted / bad input? Or do you need to add checks for `CF_BUFSIZE` overflow?